### DISCUSSION TITLE
Back End team's contributions to the Project Plan

### DISCUSSION THREAD

**Discussion by mehmetkuzu** [bounswe/bounswe2023group2#303] [General]
It will be a good practice if we reply to comments about different ideas one by one.

**Comment by mehmetkuzu** [comment]
### Define API Conventions
- Conventions that will also effect front-end/mobile code
   - Use of response codes - for success or failure
   - json style in general (a sample/template json response for different operations -data queries, data updates, logins etc)
- Conventions that will only be shared in back-end code
   - Exception handling (Use of user-defined/custom exceptions. Iterative development for system wide exception classes) 
   - Commenting/documentation standards

First 1 week of back-end work....

**Reply by mehmetkuzu** [reply]
We will not try to be perfect about this. (Like deciding on/discussing design patterns, naming conventions etc.)
But writing some template/base codes in the beginning will save us time and provide easy/good coordination.

**Comment by mehmetkuzu** [comment]
### Write a database manipulation codebase
- Some kind of adapter code for mongodb operations.
- Code for concrete data manipulation (eg. a function to update user data) This will of course be a continous effort. But defining the class structure should be a starting task. Ideally back end team members should call the same function/code for the same operation)

The starting task can be finished first week. (Together with [conventions task](https://github.com/bounswe/bounswe2023group2/discussions/303#discussioncomment-7250570))

**Comment by mehmetkuzu** [comment]
### Back-end code for Milestone 1
**This should be a decision of all teams for sure.**
- Propose these to be included in milestone 1
   - Login 
   - Registration
   - User validation/verification
   - User profiles (part of it)
- Back end coding (documented - at least inline) for these

2 weeks (after [conventions](https://github.com/bounswe/bounswe2023group2/discussions/303#discussioncomment-7250570) and [database manipulation codebase tasks](https://github.com/bounswe/bounswe2023group2/discussions/303#discussioncomment-7250682) are finished) will be ok I think.

**Reply by bgmrsln** [reply]
Shoul we also write the tests for the registration?

**Comment by bgmrsln** [comment]
Final Milestone suggestions 
Profile page design consisting of user details, user activities, subcribed notifications.
Update user activity endpoints in backend
Update user details endpoints in backend
Tests

**Comment by mehmetkuzu** [comment]
### İyi düşün Back-end team: Automated testing
:-)
(Kaşınmak gibi olmayacaksa) Actually it would be a good practice to write unit tests for endpoints. (Last season some of us did that good)
So to the project plan we can add:

- Unit test for fastapi endpoints. (This will be planned for CM1. We don't need to write a sample or template for this. Instead while coding for CM1 we can also code the unit tests. And these will be our templates for the continuing work)
- CI/CD definitions for automated tests (Also planned for CM1)
